### PR TITLE
[GeoMechanicsApplication] Remove duplicated member functions of SmallStrainUPwDiffOrderElement +

### LINKED ISSUE #12466: [GeoMechanicsApplication] Remove duplicated member functions of `SmallStrainUPwDiffOrderElement`
**Acceptance Criteria**

Member functions that duplicate the functionality as implemented by its _new_ base class (i.e. `UPwBaseElement`) have been removed.

**Implementation Details**

The following member functions have been deemed to be duplicates of the base class's ones (i.e. they are identical to the ones of class `UPwBaseElement`):
- `CalculateDampingMatrix`
- `CalculateIntegrationCoefficient`
- `CalculateIntegrationCoefficients`
- `CalculateLeftHandSide`
- `CalculateLocalSystem`
- `CalculateOnIntegrationPoints` (only the overload that accepts a `Variable<ConstitutiveLaw::Pointer>` instance)
- `CalculateRightHandSide`
- `EquationIdVector`
- `GetDofList`
- `GetFirstDerivativesVector`
- `GetIntegrationMethod`
- `GetSecondDerivativesVector`
- `ResetConstitutiveLaw`
- `SetValuesOnIntegrationPoints` (only the overloads that accept `Variable<double>` or `Variable<Matrix>` instances)

If one or more of these functions are not covered, first tests should be added, in a different issue.

### PR DESCRIPTION
**Pull request by markelov208**
**📝 Description**
The following functions have been removed from SmallStrainUPwDiffOrderElement:
- GetDofList
- ResetConstitutiveLaw
- CalculateLocalSystem
- CalculateLeftHandSide
- CalculateRightHandSide
- EquationIdVector
- GetFirstDerivativesVector
- GetSecondDerivativesVector
- CalculateDampingMatrix
- SetValuesOnIntegrationPoints (only for double and Matrix)
- CalculateOnIntegrationPoints (only for Variable<ConstitutiveLaw::Pointer> instance)

In addition
- Initialize calls the correspondent function of the base class now.
- GetDofs is virtual and protected in UPwBaseElement class and its children.
For transient_Pw_element class the following functions are removed:
- GetDofList
- EquationIdVector

Hopefully, 1d_consolidation test covers all these functions in SmallStrainUPwDiffOrderElement.

### CODE REVIEW

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp [hunk 1]
```diff
@@ -181,7 +181,7 @@ void UPwBaseElement::ResetConstitutiveLaw()
 
 void UPwBaseElement::GetDofList(DofsVectorType& rElementalDofList, const ProcessInfo&) const
 {
-    rElementalDofList = GetDofs();
+    rElementalDofList = this->GetDofs();
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: I think this is only needed if there is no definition of GetDofs in the UPwBaseClass. In other words, only if you use a protected function, that you don't have an override for in the class you are calling the function from.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.hpp [hunk 2]
```diff
@@ -173,11 +173,10 @@ class KRATOS_API(GEO_MECHANICS_APPLICATION) UPwBaseElement : public Element
 
     virtual std::size_t GetNumberOfDOF() const;
 
-    StressStatePolicy& GetStressStatePolicy() const;
+    StressStatePolicy&                   GetStressStatePolicy() const;
+    [[nodiscard]] virtual DofsVectorType GetDofs() const;
```
Comment (rfaasse) [inline applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.hpp]: For me it's not completely clear why the GetDofs function needs to be protected now, does the diff order element use it?

**Review by rfaasse — COMMENTED** [review]
This is a very nice clean up! In a next PR we can go further by moving some of the functions we have in UPwSmallStrain and SmallStrainUPwDiffOrder (e.g. for calculating the lists) to the UPw base, but that's not really part of this PR. 

I have a few minor questions left, but other than that I think it's ready to go. However, since @avdg81 worked quite a bit on the dofs and changing them around, I'd like to wait for his review before we merge.

On a side note: could you use the 'Squash and Merge' option next time? That helps in keeping the master history cleaner and more linear:
image

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp (hunk 1 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: My understanding is that `this` is required when using class templates and referring to members of base classes. That's why you see many occurrences of `this->` in e.g. `UPwSmallStrainElement`, but almost none in `SmallStrainUPwDiffOrderElement`.

For more background info, please refer to [cppreference.com](https://en.cppreference.com/w/cpp/language/this). There you will find the following quote:
> In class templates, `this` is a [dependent expression](https://en.cppreference.com/w/cpp/language/dependent_name), and explicit `this->` may be used to force another expression to become dependent.

To summarize, I agree with Richard's suggestion to remove `this->` when it is not strictly necessary.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.hpp (hunk 2 above)
Comment (avdg81) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.hpp]: Also here I agree with Richard. If `SmallStrainUPwDiffOrderElement` only provides the implementation (which I think it does), but it doesn't call this member itself, `private` should be fine.

**Review by avdg81 — COMMENTED** [review]
This cleans up the code really nicely. Thanks for doing all the work @markelov208 . Much appreciated. I agree with the comments by @rfaasse, so if you could take care of those, that would be great. Apart from that, I think this PR is good to go.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp (hunk 1 above)
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.cpp]: Thank you for the info. It is reverted and works.

File: applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.hpp (hunk 2 above)
Comment (markelov208) [reply applications/GeoMechanicsApplication/custom_elements/U_Pw_base_element.hpp]: private function also works.

**Review by markelov208 — COMMENTED** [review]
Richard and Anne, thanks a lot for the reviews and info. Concerning the commits, I thought it is useful to keep all of them. For example, Kratos web-site lists developers based on commits. Perhaps, on merged PRs. ;(

**Review by avdg81 — APPROVED** [review]
Look at how many lines you have been able to remove :-) Really nice. I believe this PR is ready to go. And please use the squash and merge option as suggested by Richard. Thank you.

**Review by rfaasse — APPROVED** [review]
Thank you for your work on this, very nice clean-up!
